### ISSUE TITLE
Runaway connection attempts with `pgxpool`

### ISSUE THREAD

**Issue by mbrancato** [jackc/pgx#1780]
**Describe the bug**
I have an application using a `pgxpool.Pool` (pgx/v4 v4.18.1) to manage connections to a PgBouncer instance, and ultimately to a PostgreSQL 15 server. 

**To Reproduce**
I am currently trying to reproduce this in a simple manner. The connection setup looks like this:

```go
	poolConfig, err := pgxpool.ParseConfig("postgres://postgres:postgres@localhost:5432/postgres?sslmode=disable")
	if err != nil {
		sugar.Fatalf("error parsing DB URL: %v", err)
	}
	poolConfig.MaxConns = 4
	poolConfig.MinConns = 4
	poolConfig.ConnConfig.Logger = zapadapter.NewLogger(logger)
	poolConfig.ConnConfig.LogLevel = pgx.LogLevelDebug

	dbPool, err := pgxpool.ConnectConfig(ctx, poolConfig)
	if err != nil {
		sugar.Fatalf("error connecting to DB: %v", err)
	}
```

We're using the `QueryRow()` etc directly on the pool and not using `Acquire()`. The resulting `Row` is passed to `Scan()` eventually.

**Expected behavior**
The pool should fill and keep 4 connections.

**Actual behavior**
The pool fills up, then for some unknown reason, it is attempting to create more connections constantly. The data from `pool.Stat()` indicates there are already 4 total connections, but continues to attempt to create more.

At the debug log level, `pgxpool` is not clearly indicating why it is creating multiple connections, but it is easy to see the attempts:
<img width="325" alt="image" src="https://github.com/jackc/pgx/assets/8493275/452f56bc-b47d-4605-8d16-f0f76708f6dd">

Note: the connection failures here are all PgBouncer running out of client connections, which has way more than pool size * number of copies of app.

```
 server error (FATAL: no more connections allowed (max_client_conn) (SQLSTATE 08P01))
```

Stat info for the above logs:
before connection runaway (`2023-10-26 05:34:14.968`):
```json
{
    "max_lifetime_destroy_count": 0,
    "constructing_connections": 0,
    "acquired_connections": 0,
    "empty_acquire_count": 0,
    "new_connections_count": 4,
    "acquire_count": 1,
    "acquire_duration": 2.6e-7,
    "max_idle_destroy_count": 0,
    "total_connections": 4,
    "canceled_acquire_count": 0,
    "message": "Database pool stats",
    "idle_connections": 4,
    "timestamp": "2023-10-26T09:34:14.968Z",
    "max_connections": 4
}
```

shortly after connection runaway (`2023-10-26 05:34:29.969`):

```json
{
    "idle_connections": 0,
    "canceled_acquire_count": 0,
    "constructing_connections": 0,
    "max_connections": 4,
    "message": "Database pool stats",
    "total_connections": 4,
    "max_lifetime_destroy_count": 0,
    "empty_acquire_count": 5,
    "timestamp": "2023-10-26T09:34:29.969Z",
    "max_idle_destroy_count": 0,
    "acquire_duration": 34.920843834,
    "acquire_count": 10,
    "acquired_connections": 4,
    "new_connections_count": 4
}
```

1 minute after connection runaway (`2023-10-26 05:35:24.975`):
```json
{
    "max_lifetime_destroy_count": 0,
    "idle_connections": 0,
    "new_connections_count": 15,
    "max_idle_destroy_count": 0,
    "empty_acquire_count": 299,
    "acquired_connections": 3,
    "acquire_count": 304,
    "message": "Database pool stats",
    "constructing_connections": 0,
    "canceled_acquire_count": 4,
    "timestamp": "2023-10-26T09:35:24.975Z",
    "acquire_duration": 2790.153294167,
    "max_connections": 4,
    "total_connections": 3
}
```

**Version**
 - Go: 1.20
 - PostgreSQL: 15 (CloudSQL), with PgBouncer 1.20.1
 - pgx: `v4.18.1`

**Additional context**
I started setting MaxConns and MinConns to mitigate this, with no luck.

**Comment by jackc** [comment]
I haven't seen anything like that before.

A few things to try:

* Connect directly to the PG server to eliminate PgBouncer as a factor
* When you have the simpler reproduction case, also try it with v5. 

---

Actually, it doesn't appear that the pool is overfilling. That would indicate that existing connections are dying. Are you canceling contexts? That terminates the underlying connection. Based on your stats only 15 total connections were made. It's quite possible that canceled queries are triggering the new connections.

**Comment by mbrancato** [comment]
Thanks for the suggestion @jackc 
There was a timeout on the context for `pgxpool.ConnectConfig(ctx, poolConfig)`. I moved it to simply `context.Background()` but it didn't seem to fix the issue. If a query gets a context timeout (context passed to `QueryRow`), then immediately after that the pool appears to attempt to create new connections.

The surprising thing is that the connection is orphaned or something when a query times out.

```json
{"severity":"ERROR","timestamp":"2023-10-27T13:30:09.268Z","message":"Query","pid":3727644839,"err":"timeout: context deadline exceeded","sql":"....","time":10.0033073,"args":["..."]}
{"severity":"ERROR","timestamp":"2023-10-27T13:30:09.270Z","message":"Query","err":"timeout: context deadline exceeded","sql":"...","time":10.000524572,"args":["..."],"pid":3765987911}
{"severity":"DEBUG","timestamp":"2023-10-27T13:30:09.271Z","message":"Connecting to database"}
{"severity":"INFO","timestamp":"2023-10-27T13:30:09.271Z","message":"Dialing PostgreSQL server","host":"..."}
{"severity":"DEBUG","timestamp":"2023-10-27T13:30:09.272Z","message":"Connecting to database"}
{"severity":"INFO","timestamp":"2023-10-27T13:30:09.273Z","message":"Dialing PostgreSQL server","host":"..."}
{"severity":"DEBUG","timestamp":"2023-10-27T13:30:09.275Z","message":"Connected to database"}
```

**Comment by mbrancato** [comment]
Looking at this, would a context cancelled or timeout cause a connection to be disconnected? And does it notify the server to disconnect. I ask the latter because if there is a graceful disconnect, I wouldn't expect to get max connection errors.

https://github.com/jackc/pgx/issues/506#issuecomment-531800089

**Comment by mbrancato** [comment]
I did notice I'm not seeing the connection `Close()` being called in the logs emitted, just constant new connection creation.

https://github.com/jackc/pgx/blob/13468eb321a0bacbb0cbdfddb1ea1d62e68e0652/conn.go#L257

We've never had an issue with v4 to date, but I might have to switch this over to v5 to see if that helps.

**Comment by jackc** [comment]
> Looking at this, would a context cancelled or timeout cause a connection to be disconnected?

Yes. 

> And does it notify the server to disconnect. I ask the latter because if there is a graceful disconnect, I wouldn't expect to get max connection errors.

It tries to cancel the running query and it sends the terminate message to the server. But it also closes the local end of the connection immediately.

**Comment by mbrancato** [comment]
hey @jackc 

I've been able to make some simple examples that show pgxpool attempting to create too many connections. As I worked on this, I started using multiple pgbouncer containers. I thought that may have been the cause due to a lack of peering in the pgbouncer replicas. Ultimately, I was able to see it happening directly against the database server.

I've created two branches to demo this, one for v4 and once for v5. Both seem to trigger the same behavior.

v4: https://github.com/mbrancato/pool-party/tree/pgx/v4/pgxpool
v5: https://github.com/mbrancato/pool-party/tree/pgx/v5/pgxpool

**Comment by jackc** [comment]
I don't have docker setup on my development machine (I'm on MacOS), but I ran the example against my local database both over TCP and local socket. I wasn't able to duplicate the error.

But it doesn't actually surprise me too much that it could occur when the application pool size is exactly the same as the server max connections. pgxpool considers the slot available for a new connection when the termination message has been sent and the local side of the network connection has been closed. But it may take the server side process a little while to terminate.

**Comment by mbrancato** [comment]
I'm using [Rancher Desktop](https://rancherdesktop.io/) as my "Docker" implementation on MacOS as well. Mainly as it provides a quick way to setup multiple things including a TCP loadbalancer and many pgbouncer copies.

I understand the example is using a small connection max count. At scale, I'm seeing these connections runaway at several times the size of the max pool size. I can also duplicate this behavior using pgbouncer in the mix, and it may be that pgbouncer does not try to discard the connection to the client until the client disconnects the socket.

The setup looks like this:
```mermaid
  graph LR;
      pgxpool-->pgbouncer;
      pgbouncer-->pgsql;
```

Where the pool / max conn values are:
pgxpool: 4
pgbouncer: 40 (max pooled client connections)
pgbouncer: 4 (actual database connections)
pgsql: 40 (max actual client connections)

I do see *some* of these errors, not many. This indicates the cancel request was sent to the wrong copy of pgbouncer. 

```
pool-party-pgbouncer-2   | 2023-10-31 02:48:13.600 UTC [1] LOG C-0x5588de338460: (nodb)/(nouser)@192.168.224.13:57658 closing because: failed cancel request (age=0s)
```

And that is likely the case because the cancel request was sent to the wrong copy. I imagine this is because the connection is not re-using an existing socket connection to the DB.

**Comment by jackc** [comment]
Hmm... I'm not sure. The cancel request has to temporarily make a new connection to issue the cancel request. Not sure if those are counted in PgBouncer or PostgreSQLs max connection counts.

I hope to release v5.5.0 today. It includes a tweak to CancelRequest for PgBouncer (5d0f904831a84cd6b7f17ef4539d2217ae3f8a66). Perhaps that will improve things.

**Comment by rkofman** [comment]
We seem to be running into a similar issue. Recently had a production issue where a single process was able to saturate DB connections (direct to DB, no pgBouncer in between), which is 20x our `MaxConns` setting.

**Comment by jlisthood** [comment]
Hi @mbrancato, thanks for raising this issue publicly and also providing a repro, we are seeing a very similar issue between pgx v4.16.1 and Pgbouncer. We have symptoms of client-side connections far exceeding expected values during periods of context cancellation.

For example, on the client-side we have a client-side connection pool limit of 100 and on the server-side we have a limit of 1024, but we are seeing this get exhausted when the only client is a sole pgxpool that is maxed out on connections.

I was reading through the old pgconn code and I noticed that query cancels are sent asynchronously:
https://github.com/jackc/pgconn/blob/831fc211bc3b06eb7b461992c231fea7496dbf29/pgconn.go#L618-L641

A question for @jackc, assuming that query cancels are generally asynchronous and are triggered when contexts passed into the various pgx interfaces are canceled - it wouldn't be too surprising that the number of client-connections from pgx to the target could exceed the max size of the connection pool? Presumably if there is some delay in processing cancellation requests, then the outbound client-connection count could be the `SUM(max_pool_size, pending_cancel_requests)`

**Comment by jackc** [comment]
> it wouldn't be too surprising that the number of client-connections from pgx to the target could exceed the max size of the connection pool? Presumably if there is some delay in processing cancellation requests, then the outbound client-connection count could be SUM(current_pool_size, pending_cancel_requests)

I'm not sure how pgbouncer or PostgreSQL itself count cancellation requests, but yes, there would be that many outbound connections.

In pgx v5, you can use pgconn.Config.BuildContextWatcherHandler to control exactly how cancellations are performed. Tuning this might solve the problem./

**Comment by james-johnston-thumbtack** [comment]
We also have been experiencing runaway client connections on PGBouncer 1.22.0 where `show pools` reports orders of magnitude more connections in `cl_waiting` status than the database/sql client connection pool size (note we are using pgx stdlib package).  Eventually this can lead to errors like `no more connections allowed (max_client_conn)`.  We did some further investigation and experimentation since we could reproduce the problem with a load test.

TL;DR of conclusions in our case:

- pgx is not leaking or otherwise holding open excessive numbers of TCP sockets to PgBouncer.  That is not the problem.
- However, the database/sql connection pool (or probably pgxpool for that matter?) will allow the app to aggressively reconnect to PgBouncer, without any sort of backoff / delay mechanism.  _**It would be beneficial if there was an easy way to configure some sort of exponential backoff when reconnecting to PostgreSQL from pgxpool.**_  For example, if queries are failing due to transient reasons (e.g. canceled contexts) and the connection is being discarded, then creation of new connections should start to be throttled.  We think this might be a way that would allow overwhelmed PGBouncer instances to recover.

Example PGBouncer output in our case (note that cancel requests are tracked separately):
```
pgbouncer=# show pools;
 database  |        user        | cl_active | cl_waiting | cl_active_cancel_req | cl_waiting_cancel_req | sv_active | sv_active_cancel | sv_being_canceled | sv_idle | sv_used | sv_tested | sv_login | maxwait | maxwait_us |  pool_mode
-----------+--------------------+-----------+------------+----------------------+-----------------------+-----------+------------------+-------------------+---------+---------+-----------+----------+---------+------------+-------------
 pgbouncer | pgbouncer          |         1 |          0 |                    0 |                     0 |         0 |                0 |                 0 |       0 |       0 |         0 |        0 |       0 |          0 | statement
 users     | users_latency_read |         0 |       2027 |                    0 |                     0 |         0 |                0 |                 0 |       0 |       0 |         0 |        1 |      10 |      52536 | transaction
```
In this case, 10 load test programs were running, each with a connection pool size of 30, so we should never see more than 10 * 30 = 300 connections across `cl_active + cl_waiting`.  But here, `cl_waiting` is 2027.  Note that the PGBouncer instance has exhausted all server connections: we find it's in a loop of logging in and then immediately discarding the connection, over and over (in our case it's because the client is timing out and cancels the Go context passed to QueryContext).  Thus PGBouncer is in a bad state, and it won't recover unless you make it shed the load.

Given the discussion in this issue around `asyncClose` which has a fire-and-forget Goroutine in it, which could hold TCP connections open that are no longer considered part of the connection pool (and thus subject to pool limits), we questioned whether this could be the excess of connections shown here.  We temporarily changed `go func() {` to just `func() {` in `asyncClose`.  Also we temporarily added a global counter to track how many concurrent `asyncClose` functions were running, and it was never very high.  Our load test could still cause the above condition.  _**Therefore, in our case, we conclude that the goroutine in the `asyncClose` function in pgx is not the cause of the "runaway connection counts" on PgBouncer.**_

This was further proved by examining `netstat` output.  _**What we found is that pgx is dutifully closing the sockets such that the number of open sockets in `ESTABLISHED` state is never more than the connection pool size times two (doubled for the cancel requests).**_  However, when PGBouncer shows runaway state above, we find a large number of additional `netstat` connections on the client in `FIN_WAIT2` state.  (On the PGBouncer server, they show as `CLOSE_WAIT` state).  _**This means that the client (pgx) has closed the TCP socket, but the server application (PGBouncer) has yet to finish working with the socket on its end and close it as well - PGBouncer still has the PGBouncer server socket held open for lengthy periods of time, long after pgx has closed pgx's socket.**_

What we are currently concluding is that the single-threaded nature of PGBouncer may be a limitation.  The hypothesis is that there is this timeline:

1. Client app (pgx) connects to PGBouncer.  PGBouncer immediately answers.
2. Client app asks to connect to app database and sends the PG credentials.
3. Client app asks to run a query.
4. PGBouncer puts the connection into waiting state, because there is not a PG database connection available.
5. A few seconds pass or whatever because PGBouncer has 2000 other connections in front of it that are waiting.
6. The client gives up and the context is canceled on the client.
7. Go code decides to throw out the connection since we canceled context and the state of the connection is poisoned.  The connection is promptly closed and the pgx socket is closed and the connection moves to `FIN_WAIT2` state on the client and `CLOSE_WAIT` on the server.
8. Yet another few seconds later - finally PGBouncer is ready to serve the connection!  (Remember it is only making one connection attempt to a PG server at a time, so it took awhile to get to us).  It logs into PG and associates the new PG connection with the client connection.  (Note this is proved by polling `pg_stat_activity` on the reader PG instance: the list is empty apart from an occasional fleeting connection if you poll it at the right time.  You can see `sv_login` appears stuck as 1.)
9. Uh oh - PGBouncer finally realizes that the client has disconnected!  (In reality the client did that a long time ago in step 7).  PGBouncer immediately disconnects from PostgreSQL since the connection is poisoned.  So all that hard work it went to, waiting to establish a new PG connection?  It all gets thrown away now.
10. Now PGBouncer will close the client connection socket, and finally the operating system closes out the connection completely.  In the mean time, the client application had failed, and more connections stack up in PGBouncer, enqueued/waiting connections from all the client's retry attempts.

Some things that we think may help (we are still investigating/experimenting):
- Reducing `query_wait_timeout` in PGBouncer to the client application's timeout.  When the client application has a canceled context and pgx closes the socket, it's useless for PGBouncer to leave the connection in the cl_waiting queue.  Our application has 2 second timeout, after which the context will end up canceled.  So we set `query_wait_timeout` to 2, and we find the number of connections in cl_waiting state will decrease significantly (but alone it is not enough to get PgBouncer out of the bad state).
- Horizontally scaling PgBouncer.  PgBouncer is single threaded and can only initiate one new server connection at a time, so more PgBouncer instances means more threads available to start new server connections.
- _**Exponential backoff / throttling in pgx client connection pools when connections are failing.**_  This is an area where pgx could help.  Is there a good way to delay creating new connections in pgxpool if queries are failing due to transient reasons?  We think this could help temporarily reduce load to PgBouncer and allow it to recover.

**Comment by james-johnston-thumbtack** [comment]
Follow up on this, here's what has been found helpful, and makes a big difference in a load test:

**PgBouncer changes:**

- Reducing `query_wait_timeout`: this definitely helps PgBouncer shed the overwhelming load faster and help get it unstuck.  Just reducing this parameter measurably has an immediate effect on the number of pending connections in `cl_waiting` state, and thus makes it more likely that PgBouncer will recover on its own.  If your client code has Go contexts which are being canceled, you can set this value to be no higher than the context timeouts - there's no reason for PgBouncer to keep waiting to serve client connections that have canceled contexts.
- Horizontally scaling the single-threaded PgBouncer that can only make one login attempt at a time, probably helped as well.

However, although they helped, these alone were not enough to totally resolve the problem.  I think they are complementary to the client changes below.

**Client changes:**

We have services with mixtures of either standard database/sql pools using pgx/stdlib (in older code), and/or directly using pgxpool (in newer code).  If you are still using the standard database/sql pool (i.e. using `sql.Open` with the pgx/v5 driver), then I would suggest to first switch to using pgxpool as an initial step.  This is because pgxpool provides more advanced hooking capabilities, which we will need.  Look into directly constructing the pgxpool, and then use [stdlib.OpenDBFromPool](https://pkg.go.dev/github.com/jackc/pgx/v5@v5.7.5/stdlib#OpenDBFromPool) to wrap it.  Beware that both pools have to be closed in the correct order: first `sql.DB.Close()` and then close the `pgxpool.Pool`.  Also beware that you must not manipulate the returned database/sql.DB pool settings with the `Set*` functions (e.g. `SetMaxOpenConns`) because it is a wrapper for the pgxpool - instead, you need to change code to directly configure the pgxpool (e.g. using the connection string options that pgxpool recognizes).

Next, we tried out [adaptive throttling as described in Google's SRE handbook](https://sre.google/sre-book/handling-overload/) and this made a big difference.  The implementation from https://pkg.go.dev/github.com/bradenaw/backpressure seems to work well enough...

To tie it together with pgxpool.PoolConfig, we created a Tracer implementation, and a BeforeConnect hook:
- We created two independent adaptive throttlers from `backpressure`: one for new connection failures, and another one for when connections fail to be returned to the pool.  We found this might be best/easier than a combined throttler, since some failure modes with PgBouncer seem to result in enough successful connection attempts to outweigh the subsequent canceled contexts during a query.
- The Tracer implements `pgx.ConnectTracer`.  In `TraceConnectEnd`, we call `Rejected` or `Accepted` on the adaptive throttler for new connection failures based on if there was an error in `data.Err`.
- The Tracer also implements `pgxpool.ReleaseTracer`.  Here, we copy/pasted the conditional from https://github.com/jackc/pgx/blob/v5.7.5/pgxpool/conn.go#L33 to predict whether the connection would be returned to the pool, or destroyed because the connection was in an erroneous state (e.g. query still running because a context was canceled during it, or some other cause).  We instruct the adaptive throttler for when connections fail to be returned to the pool, with `Rejected` if the connection will be destroyed, and `Accepted` if it will be returned to the pool.
- In the BeforeConnect hook, we sequentially call the `Attempt` function on first the throttler for new connection failures, and then the throttler for connection failed to be released to the pool.  If the `Attempt` function indicates we should throttle, then BeforeConnect will return a throttled error.

With this implementation, we found that when PgBouncer was overloaded during a load test, there may still be some temporary periods of time where `SHOW POOLS` would report `cl_waiting` > 0 and `cl_active` would be low values (even 0), but when this would happen, it would eventually tend to stabilize with everything in `cl_active` and nothing waiting.  This was a dramatic improvement over the previously observed behavior, where once PgBouncer was overwhelmed, it was down until you dropped almost all the load testing traffic.

**Suggested changes for pgx**

**The first obvious change that would be a little better would be if `pgxpool.TraceReleaseData` struct contained more information about the pending release.  In particular, it would be useful to know if the released connection will be returned to the pool, or destroyed.  And, if it will be destroyed, then to know the reason why.**  Our implementation had to copy/paste the obscure `if conn.IsClosed() || conn.PgConn().IsBusy() || conn.PgConn().TxStatus() != 'I'` condition from pgx source code into our tracer, which is obviously not ideal.  (For example, we would call `Rejected` on the adaptive throttler if the connection is still busy, but we would call `Accepted` if the connection is closed simply because it reached max configured lifetime, since the latter is not a reason why we should consider throttling.)

Beyond that, I am not sure if any more substantial changes are truly needed in pgx, because the hooking capabilities were good enough to integrate with the `backpressure` package with minimal code that didn't feel too hacky (other than the previous paragraph).  pgx could consider directly implementing this integration so that adaptive throttling parameters could, for example, be provided as part of the pool connection string: this would still save a fair bit of time and boilerplate out of the box.  Of course, users may want a variety of schemes - maybe one person wants the fail-fast adaptive throttling mechanism we implemented above, but maybe a different user would want an adaptive rate limiter where `BeforeConnect` simply blocks for a period of time rather than immediately returns a throttled error, and yet another user might want a simple non-adaptive static rate limiter in `BeforeConnect` without using any feedback from the tracers at all.

**Comment by jackc** [comment]
> The first obvious change that would be a little better would be if pgxpool.TraceReleaseData struct contained more information about the pending release. In particular, it would be useful to know if the released connection will be returned to the pool, or destroyed. And, if it will be destroyed, then to know the reason why.

I'm open to this.

---

One other thing that may help is to provide a custom `BuildContextWatcherHandler`. It lets you entirely customize cancellation behavior. You can change if and when to set deadlines on the Go side. You can choose whether to send a CancelRequest to the server or not. Turning that off would save all the new connections created just to cancel an existing request. If you want to get really fancy, you could have a dedicated connection that called `pg_cancel_backend()`, though I'm not sure how that would work with PgBouncer in the way.

**Comment by james-johnston-thumbtack** [comment]
I haven't investigated, but I'm pretty sure cancel requests are probably thoroughly broken in our current setup.  But I'm also a little skeptical it was the cause of our problems.  We have many PgBouncer instances behind a load balancer.  So by default, a cancel request will probably not go to the same PgBouncer instance that was proxying the original query (and thus, to the same PG back-end).  PgBouncer [has a whole "peering" section of configuration](https://www.pgbouncer.org/config.html#section-peers) to help it route cancellation requests to the right place, but we never set it up.  So I'm guessing most of our cancel requests aren't actually doing anything.  Presumably most of the cancellation requests are going to PgBouncer instances that don't know anything about the ongoing query, and thus probably ignore them.  We do have 1 second statement timeouts in most cases, so we don't really have a big probably with abandoned PG back-end workers running amok for extended periods of time.

I'm more suspicious that our problem has to do with the main TCP connections closing.  E.g. pgx gets the context cancellation, and pgx closes the TCP connection because the connection is in an indeterminate state (e.g. what if you were in the middle of downloading a result set, or you were mid-transaction, etc.?).  You don't want to put a connection back in the pool unless you know it's safe to use again and wasn't in the middle of something.  Similarly, PgBouncer would need to have a similar semantic: if the client disconnects in the middle of a statement, then the back-end server connection could have been in the middle of sending a result set back to the client, etc. and the back-end server needs to disconnect.  In this case, I think PgBouncer lacks a bit of smarts and doesn't realize it doesn't have to invalidate a fresh server connection when the client connection already disconnected by the time the server connection is allocated to a waiting client.

So presumably, if a custom `BuildContextWatcherHandler` were to help in this scenario, I'd need it to _not_ disconnect right away from PgBouncer, and instead finish out the current statement as fast and cleanly as possible.  For example, wait for query to finish or reach statement timeout, and then cleanly close the cursor without actually reading any rows, thus leaving the connection in a good state again before pgx disconnects from PgBouncer, so that PgBouncer doesn't feel a need to discard the server connection.  I have no idea if this is actually feasible to do, since after all, the Go code that called pgx that would have read from the result set is no longer going to do so.

Caveat is I am by no means any kind of expert on the PG wire protocol, so I could be a little wrong in all the above.  I'm not sure if my idea for `BuildContextWatcherHandler` makes a lot of sense in the first place, or if it's possible to write one with pgx that would finish waiting for the query and then close the cursor so that the server connection is ready for another statement to be issued (and thus could be reused by PgBouncer).

**Comment by jackc** [comment]
> I haven't investigated, but I'm pretty sure cancel requests are probably thoroughly broken in our current setup.

If that's the case then I'd suggest using different `BuildContextWatcherHandler` that only sets a net.Conn deadline and doesn't also send the `CancelRequest`. See `DeadlineContextWatcherHandler`.

> But I'm also a little skeptical it was the cause of our problems.

Removing the cancellation requests would presumably reduce the load on PgBouncer.

>  if it's possible to write one with pgx that would finish waiting for the query and then close the cursor so that the server connection is ready for another statement to be issued (and thus could be reused by PgBouncer).

Unfortunately, there isn't a way to do this at least at the moment. A long time ago pgx tried to do this, but there were edge cases that were very difficult to reliably handle, so I ended up going the other way and defaulting to a network deadline that when triggered closed the connection.

That said, depending on your exact needs `CancelRequestContextWatcherHandler` can often suffice. That allows you to control a delay after the context is canceled before the deadline is passed and a separate delay for how long before a `CancelRequest` is sent. By default pgx prioritizes returning from a method where the context has been canceled as quickly as possible. But if you are willing to wait a little while, you can drastically reduce the amount of connections that must be closed.

For example, assuming `CancelRequest` was working properly, you might set a 0 delay on sending the `CancelRequest`, but a 500ms or 1000ms delay on the deadline. In most cases, the query would get the canceled request error before the deadline expired. That would mean the connection stays alive.

**Comment by jlisthood** [comment]
>We have many PgBouncer instances behind a load balancer. So by default, a cancel request will probably not go to the same PgBouncer instance that was proxying the original query (and thus, to the same PG back-end).

We have the same problem. Cancel requests very rarely work in our setup for the same reasons. We are going to disable them on the client-side. It does not appear to me that there are major downsides to this, except that some cancel requests that were working will not work now. But given how low the success rate must be, I don't think it is worth the extra overhead on the single-threaded pgbouncer to process these cancel requests.

I'd love for query cancellations to work, but that is probably a discussion for the pgbouncer community. Peering seems promising, but complicated to keep the peering configs accurate in some setups.

> To tie it together with pgxpool.PoolConfig, we created a Tracer implementation, and a BeforeConnect hook:

Thanks for sharing the details on this. We will explore the same solution. One limitation is that we are still using pgx v4, so we need to either upgrade to v5 to take advantage of the new tracers available or patch v4 internally. This is good motivation to do the upgrade though.

**Comment by james-johnston-thumbtack** [comment]
We have subsequently found that the adaptive throttling approach I described in my [previous message](https://github.com/jackc/pgx/issues/1780#issuecomment-3085650843) was not enough to fully mitigate the problem.  What _has_ helped considerably is adapting @jackc 's [suggestions on setting `BuildContextWatcherHandler`](https://github.com/jackc/pgx/issues/1780#issuecomment-3146833352).  This pool configuration made a huge difference:

```
	poolConfig.ConnConfig.BuildContextWatcherHandler = func(
		pgConn *pgconn.PgConn,
	) ctxwatch.Handler {
		return &pgconn.DeadlineContextWatcherHandler{
			Conn:          pgConn.Conn(),
			DeadlineDelay: 10 * time.Second,
		}
	}
```

This effectively allows our app to ignore the canceled context long enough finish whatever it was doing with the current PostgreSQL statement (up to 10 seconds), and thus presumably allow the connection to be returned to the connection pool in PgBouncer (and probably pgx), and therefore avoid the failure mode I described earlier.

Since we do not have peering set up correctly for PgBouncer to route cancellation requests to PostgreSQL, we do not bother with `CancelRequestContextWatcherHandler` and instead use the simpler `DeadlineContextWatcherHandler`.  Most of our roles have statement timeouts of 1 second configured, so there's reduced need to have cancel requests working correctly.

**Comment by nikita17th** [comment]
> Follow up on this, here's what has been found helpful, and makes a big difference in a load test:
> 
> **PgBouncer changes:**
> 
> * Reducing `query_wait_timeout`: this definitely helps PgBouncer shed the overwhelming load faster and help get it unstuck.  Just reducing this parameter measurably has an immediate effect on the number of pending connections in `cl_waiting` state, and thus makes it more likely that PgBouncer will recover on its own.  If your client code has Go contexts which are being canceled, you can set this value to be no higher than the context timeouts - there's no reason for PgBouncer to keep waiting to serve client connections that have canceled contexts.
> * Horizontally scaling the single-threaded PgBouncer that can only make one login attempt at a time, probably helped as well.
> 
> However, although they helped, these alone were not enough to totally resolve the problem. I think they are complementary to the client changes below.

Hello, we're experiencing similar symptoms. Have you tried setting min_pool_size to a non-zero value? As far as I understand, this can reduce the number of connections pgbouncer needs to open to the Postgres server during peak loads?
